### ISSUE TITLE
QEMU register state mismatch

### ISSUE THREAD

**Issue by BugraEryilmaz** [parsa-epfl/flexus#50]
Description
There are some instructions that trigger incorrect register state after committing. 

Steps to Reproduce
Run the timing model with any workload.

Expected Behavior
The register state of QEMU and KnottyKraken should match after every commit.

Actual Behavior
```
60149 <cycle.cpp:1301> {28670485066}- [0;32mretire Move instruction to the secondary retirement buffer #47[00]  {clsComputation}  QEMU disas: add x21, sp, #0x150 {raised}	{retired}[0m
60150 <cycle.cpp:1465> {28670485066}- 000-uarch FinalCommit:#45[00]  {clsStore}  QEMU disas: stp x26, x27, [sp, #0xd0] {raised}	{retired}
60151 <Effects.cpp:1378> {28670485066}- CPU[00]#45 CommitStoreEffect 
60152 <ValueTracker.hpp:384> {28670485066}- CPU[0] Store p:100f26920[QuadWord] = e0000101f31fc3ffffff80c0f1f988
60153 <ValueTracker.hpp:407> {28670485066}- CPU[0] Store.New p:100f26920 no prior outstanding values. Now:  G:40000000000041ffffff80c0e8df80 00:e0000101f31fc3ffffff80c0f1f988<1>. Simics Reflects: 0
60154 <Effects.cpp:1382> {28670485066}- #45[00]  {clsStore}  QEMU disas: stp x26, x27, [sp, #0xd0] {raised}	{retired} Commit store instruction
60155 <cycle.cpp:1468> {28670485066}- 000-uarch commit effects complete
60156 <cycle.cpp:1498> {28670485066}- [0;32mcommit #45[00]  {clsStore}  QEMU disas: stp x26, x27, [sp, #0xd0] {raised}	{retired}[0m
60157 <cycle.cpp:1504> {28670485066}- [0;32mcommit Instruction is neither annuled nor is a micro-op[0m
60158 <cycle.cpp:1507> {28670485066}- Pre Validating... 1
60159 <microArch.cpp:376> {28670485066}- [0;32madvance[0m
60160 <accounting.cpp:311> {28670485066}- Get a Store..........
60161 <cycle.cpp:97> {28670485066}- Trying to [VALIDATE] Flexus CORE against QEMU
60162 <cycle.cpp:107> {28670485066}- Register 21 mismatch: QEMU= 0x400000000000001 Flexus= 0xffffffc080fbb9a0
```

**Comment by BugraEryilmaz** [comment]
Problem 1: register mapping effects were done in retirement not in commit which shows mismatch because we see the effects of instructions that are not committed yet. 
Fixed by [406eae51128dfa780221e0246a257ad2430895d1](https://github.com/parsa-epfl/flexus/commit/406eae51128dfa780221e0246a257ad2430895d1)

Problem 2: The system registers are not synchronized and kept up to date. 
Fixed by [0ca7b36350385ca78bb3358b151795c3abab986a](https://github.com/parsa-epfl/flexus/commit/0ca7b36350385ca78bb3358b151795c3abab986a), [5e8765d62d4efe7ff0282bfe065fa2498cff27e7](https://github.com/parsa-epfl/flexus/commit/5e8765d62d4efe7ff0282bfe065fa2498cff27e7), [5de6ec4455c85ff700b865ff4960bf9ea67693a4](https://github.com/parsa-epfl/flexus/commit/5de6ec4455c85ff700b865ff4960bf9ea67693a4), [47e82ee46a343d0b8c5ae8ee0c496e45a4e02189](https://github.com/parsa-epfl/flexus/commit/47e82ee46a343d0b8c5ae8ee0c496e45a4e02189)

Problem 3: Some system registers have a corner case (for some reason). For example, reading SP_EL1 does not always return the stack pointer of exception level 1. If we are in exception level 2 and the field HCR_EL2.E2H is 1, then SP_EL1 returns the stack pointer of exception level 2. There are a couple more edge cases here, but they have not been implemented because I did not understand what exactly they meant and did not appear in the execution. 
Fixed by [88d357ccbead8aeb7c6abb68589fedd2dd1a9736](https://github.com/parsa-epfl/flexus/commit/47e82ee46a343d0b8c5ae8ee0c496e45a4e02189), [31ba74590e16794ce8e7f3ac3773d743f0713cfd](https://github.com/parsa-epfl/flexus/commit/31ba74590e16794ce8e7f3ac3773d743f0713cfd)

All fixes are in branch fix/sysRegs

**Comment by BugraEryilmaz** [comment]
There still are some issues with `hint 0x19` and `hint 0x1d` instructions.
```
76690 <cycle.cpp:107> {28670485291}- Register 30 mismatch: QEMU= 0xd5a9c240800112e8 Flexus= 0xffffffc0800112e8
76691 <cycle.cpp:1570> {28670485291}- Post Validating... 0
76692 <cycle.cpp:1573> {28670485291}- Failed Validated #67[00]  {clsComputation}  QEMU disas: hint #0x19 {raised}	{retired}
```
According to ARM docs, if the pointer validation fails, it should set the high bits of register 30. So QEMU thinks it passes the pointer validation, but Flexus thinks it does not.
